core/include/cubos/core/net/tcp_stream.hpp

```diff
@@ -0,0 +1,82 @@
+/// @file
+/// @brief Class @ref cubos::core::net::TcpStream.
+/// @ingroup core-net
+
+#pragma once
+
+#include <cubos/core/api.hpp>
+#include <cubos/core/memory/stream.hpp>
+#include <cubos/core/net/address.hpp>
+#include <cubos/core/net/utils.hpp>
+
+namespace cubos::core::net
+{
+    /// @brief Stream implementation which writes to/reads from a TCP connection.
+    /// @ingroup core-net
+    class CUBOS_CORE_API TcpStream : public memory::Stream
+    {
+    public:
+        /// @brief Constructs an empty TCP stream, ready for connecting.
+        TcpStream();
+
+        /// @brief Deconstructs.
+        ~TcpStream() override;
+
+        /// @brief Forbid copy construction.
+        TcpStream(const TcpStream&) = delete;
+
+        /// @brief Forbid copy assignment.
+        TcpStream& operator=(const TcpStream&) = delete;
+
+        /// @brief Connects to a TCP listener, possibly on a remote machine.
+        /// @param address Destination address.
+        /// @param port Destination port.
+        /// @return Whether connecting was successful.
+        bool connect(const Address& address, uint16_t port);
+
+        /// @brief Replaces the inner socket with the given one.
+        /// @param socket Socket.
+        /// @return Whether connecting was successful.
+        void inner(InnerSocket socket);
+
+        /// @brief Disconnects inner socket.
+        void disconnect();
+
+        /// @brief Checks if inner socket is connected.
+        /// @return Whether the buffer holds a valid connection.
+        inline bool valid() const
+        {
+            return mSock != InnerInvalidSocket;
+        }
+
+        // Stream method implementations.
+
+        std::size_t read(void* buffer, std::size_t size) override;
+
+        std::size_t write(const void* buffer, std::size_t size) override;
+
+        inline std::size_t tell() const override
+        {
+            return mPos;
+        }
+
+        inline bool eof() const override
```

Comment:
Inline is redundant, you can remove the keyword - it is automatically added by the compiler if you're defining a method in the class definition itself.